File: minecraft/service.ts
import {pathfinder} from "mineflayer-pathfinder";
import {createBot} from "mineflayer";
import {plugin as autoeat} from "mineflayer-auto-eat"
import "mineflayer"
import {fertilize, harvest, sow} from "./skill/farm"
import {plugin as pvp} from "mineflayer-pvp";
import {findNearestPlayer, moveToPos} from "./util";
import {attackMobs} from "./skill/attack";
import {faceMe, followMe, wander} from "./skill/follow";
import {botHurt, botInterrupt} from "./body";
import {propitiate, tickCheckAngry} from "./brain/angry"
import {emitRespawnEvent} from "./event";
import {Entity} from "prismarine-entity";
import {fightingWithHostiles} from "./brain/intent";

const options = {
    host: process.argv[2],
    port: parseInt(process.argv[3]),
    username: process.argv[4],
    password: process.argv[5]
}

const bot = createBot(options)

console.log(`玩家 ${options.username} 成功登录 ${options.host}:${options.port}`)

bot.loadPlugin(pathfinder)
bot.loadPlugin(pvp)
bot.loadPlugin(autoeat)

bot.once('spawn', async () => {
    // @ts-ignore
    bot.autoEat.options = {
        priority: 'foodPoints',
        startAt: 14,
        bannedFood: []
    }
})

bot.on('respawn', async () => {
    await emitRespawnEvent(bot)
})

bot.on('autoeat_started', () => {
})

bot.on('autoeat_finished', () => {
})

bot.on('health', () => {
    if (bot.food === 20) bot.autoEat.disable()
    // Disable the plugin if the bot is at 20 food points
    else bot.autoEat.enable() // Else enable the plugin again
})

bot._client.on('damage_event', async (packet) => {
    const entityId = packet.entityId
    const sourceTypeId = packet.sourceTypeId
    const sourceCauseId = packet.sourceCauseId
    const sourceDirectId = packet.sourceDirectId
    botHurt(bot, entityId, sourceTypeId, sourceCauseId, sourceDirectId)
    botInterrupt(bot, entityId, sourceTypeId, sourceCauseId, sourceDirectId)
})

// @ts-ignore
bot.on("stoppedAttacking", () => {
    fightingWithHostiles.clear()
    const nearestPlayer = findNearestPlayer(bot, 5, 50)
    if (nearestPlayer) {
        moveToPos(bot, nearestPlayer.position)
    }
})
var fun = 0
bot.on('physicsTick', async () => {
    await attackMobs(bot)
    fun++;
    if (fun % 20 == 0) {
        followMe(bot)
    }
    await tickCheckAngry(bot)
})

bot.on('chat', async (username, message, translate, jsonMsg, matches) => {
    if (['来', 'lai', 'come'].includes(message)) {
        followMe(bot)
    } else if (['种', 'zhong', 'sow'].includes(message)) {
        await sow(bot)
    } else if (['收', 'shou', 'harvest'].includes(message)) {
        await harvest(bot)
    } else if (['施肥', 'shifei', 'fertilize'].includes(message)) {
        await fertilize(bot)
    } else if (['玩', 'wan', 'play'].includes(message)) {
        await wander(bot)
    }
})

bot.on('diggingAborted', (block) => {
    block.position
})

bot.on('hardcodedSoundEffectHeard', async (soundId, soundCategory, position, volume, pitch) => {
    await faceMe(bot, position, soundCategory)
})

// @ts-ignore
bot.on('attackedTarget', () => {
    propitiate(30)
})


// @ts-ignore
bot.on('blockBreakProgressEnd', async (block: Block, entity: Entity) => {
    // console.log(entity.position)
    if (entity.type === 'player') {
        moveToPos(bot, block.position.offset(0, 1, 0))
        await bot.lookAt(entity.position)
    }
})

File: minecraft/py/service.py
import json
from dataclasses import asdict
from typing import List

from javascript import require, On, Once

from minecraft.py.body import bot_hurt, bot_death, check_health_changed, check_food_changed, reset_body_info
from minecraft.py.common import create_game_event, GameEvent
from minecraft.py.eater import auto_eat
from minecraft.py.guard import attack_mobs

mineflayer = require('mineflayer', 'latest')
autoeat = require('mineflayer-auto-eat')
pvp = require('mineflayer-pvp')
pathfinder = require('mineflayer-pathfinder')
Vec3 = require('vec3').Vec3

MINECRAFT_SERVER_HOST = 'localhost'
MINECRAFT_SERVER_PORT = 25565
MINECRAFT_BOT_NAME = 'Koneko'

bot = mineflayer.createBot({
    "host": MINECRAFT_SERVER_HOST,
    "port": MINECRAFT_SERVER_PORT,
    "username": MINECRAFT_BOT_NAME,
})
# 自动进食
bot.loadPlugin(autoeat.plugin)
# PVP 插件
bot.loadPlugin(pathfinder.pathfinder)
bot.loadPlugin(pvp.plugin)

game_event_list: List[GameEvent] = []


def bot_log(msg):
    bot.chat(msg)


def select():
    if len(game_event_list) > 0:
        event = game_event_list[-1]
        if not event.read:
            event.read = True
            return event
    return None


def add_event(event: GameEvent):
    if event:
        game_event_list.append(event)
        bot_log(f'游戏事件 {json.dumps(obj=asdict(event), indent=4, ensure_ascii=False)}')


@Once(bot, 'spawn')
def handle_once_spawn(this):
    bot.autoEat.options = {
        "priority": 'foodPoints',
        "startAt": 14,
        "bannedFood": []
    }


@On(bot, 'physicsTick')
def handle_in_physics_ticks(this):
    try:
        add_event(check_health_changed(bot))
        add_event(check_food_changed(bot))
        attack_mobs(bot)
    except Exception as e:
        print(e)


@On(bot, 'spawn')
def handle_spawn(this):
    reset_body_info()
    add_event(create_game_event(bot=bot, env='你重生了'))


@On(bot, "chat")
def handle(this, username, message, *args):
    ...


@On(bot, 'health')
def handle_health(this):
    auto_eat_event = auto_eat(bot)
    add_event(auto_eat_event)


@On(bot._client, 'damage_event')
def handle_damage_event(this, packet, *args):
    entity_id, source_type_id, source_cause_id, source_direct_id = packet.values()
    bot_hurt_event = bot_hurt(bot, entity_id)
    add_event(bot_hurt_event)


@On(bot, 'death')
def death_handle(this, *args):
    bot_death_event = bot_death(bot)
    add_event(bot_death_event)
